crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -150,14 +149,96 @@ impl Acir {
+            Operation::Not(_) => todo!(),
+            Operation::Cast(value) => self.substitute(*value, evaluator, ctx),
+            i @ Operation::Jne(..)
+            | i @ Operation::Jeq(..)
+            | i @ Operation::Jmp(_)
+            | i @ Operation::Phi { .. }
+            | i @ Operation::Results { .. } => {
+                unreachable!("Invalid instruction: {:?}", i);
+            }
+            Operation::Truncate {
+                value,
+                bit_size,
+                max_bit_size,
+            } => {
+                let value = self.substitute(*value, evaluator, ctx);
+                InternalVar::from(evaluate_truncate(value, *bit_size, *max_bit_size, evaluator))
+            }
+            Operation::Intrinsic(opcode, args) => {
+                InternalVar::from(self.evaluate_opcode(ins.id, *opcode, &args, ctx, evaluator))
+            }
+            Operation::Call(..) => unreachable!("call instruction should have been inlined"),
+            Operation::Return(_) => todo!(), //return from main
+            Operation::Nop => InternalVar::default(),
+            Operation::Load { array, index } => {
+                //retrieves the value from the map if address is known at compile time:
+                //address = l_c and should be constant
+                let index = self.substitute(*index, evaluator, ctx);
+                if let Some(index) = index.to_const() {
+                    let address = mem::Memory::as_u32(index);
+                    if self.memory_map.contains_key(&address) {
+                        InternalVar::from(self.memory_map[&address].expression.clone())
+                    } else {
+                        //if not found, then it must be a witness (else it is non-initialised memory)
+                        let mem_array = &ctx.mem.arrays[*array as usize];
+                        let index = (address - mem_array.adr) as usize;
+                        if mem_array.values.len() > index {
+                            mem_array.values[index].clone()
+                        } else {
+                            InternalVar::from(self.memory_witness[&array][index])
+                        }
+                    }
+                } else {
+                    todo!("dynamic arrays are not implemented yet");
+                }
+            }
+
+            Operation::Store {
+                array: _,
+                index,
+                value,
+            } => {
+                //maps the address to the rhs if address is known at compile time
+                let index = self.substitute(*index, evaluator, ctx);
+                let value = self.substitute(*value, evaluator, ctx);
+
+                if let Some(index) = index.to_const() {
+                    let address = mem::Memory::as_u32(index);
+                    self.memory_map.insert(address, value);
+                    dbg!(&self.memory_map);
```

Comment:
can we remove this debug?

Reply:
I tried to stay as close to the original semantics as possible, but yes I can also remove dbg statements along the way if desired. I'll push a fix